### ISSUE TITLE
paywall configuration lemonde.fr

### ISSUE THREAD

**Issue by etiess** [wallabag/wallabag#2909]
I'd like to configure the access to lemonde.fr with the paywall. I tried with this version of my 'wallabag/vendor/j0k3r/graby-site-config/lemonde.fr.txt' but without success (wallabag grabs only the "free" part of the article.

I've got valid credentials to test if needed.

```
title: //h1

# We can have multiple authors
author: //a[@class='auteur']

# Last edition date (if any)
date: //time[@itemprop='dateModified']/@datetime

# Publication date
date: //time[@itemprop='datePublished']/@datetime

body: //div[@id='articleBody']

requires_login: yes
login_uri: https://secure.lemonde.fr/sfuser/connexion
login_username_field: connection[mail]
login_password_field: connection[password]
not_logged_in_xpath: //body[@class="not-logged-in"]

# Remove the insane "conjugaison.lemonde.fr" links:
find_string: <a target='_blank' onclick='return false;' class='lien_interne conjug'

replace_string: <input type='hidden' style='display:none;'

prune: no

test_url: http://www.lemonde.fr/economie/article/2011/07/05/moody-s-abaisse-la-note-du-portugal-de-quatre-crans_1545237_3234.html
```

**Comment by nicosomb** [comment]
Did you well configure your `app/config/parameters.yml` file?

**Comment by etiess** [comment]
I think so, it looks like that:

    sites_credentials:
        mediapart.fr: {username: "toto", password: "PA$$word"}
        lemonde.fr: {username:"toto@mail.com", password: "PA$$word"}

**Comment by j0k3r** [comment]
Did you remove your cache after the change?

````
php bin/console cache:clear -e=prod
```

**Comment by etiess** [comment]
yes, I did :(

**Comment by nicosomb** [comment]
@bdunogier any idea? 
@etiess I think it'll be useful to share your credentials (not here ;-) )

**Comment by j0k3r** [comment]
Could you please turn debug log on? (edit `app/config/config_prod.yml`, replace the value `error` by `debug`, clear your cache, retry and send us every line starting with `graby.` from the file `var/logs/prod.log`)

☝️ This should be in the documentation about debug paywall configuration 🙂

**Comment by etiess** [comment]
You're right ;) I propose thes changes in the documentation:
https://github.com/wallabag/wallabag/compare/master...etiess:patch-2

**Comment by etiess** [comment]
Hello, were you able to find a solution with the credentials I sent you?

**Comment by nicosomb** [comment]
@bdunogier could have a look please? I didn't find what the problem was.

**Comment by techexo** [comment]
Hi everyone. I have made some tests today because I encountered the same issues. I made the exact same modifications that @etiess made to the j0k3r graby-site-conf for lemonde.fr.

Using a paywall protected Le Monde article as an example, if I enter the URL 

> http://**www**.lemonde.fr/idees/article/2017/04/17/un_article_protege.html

, Wallabag doesn't use the credentials in parameters.yml. However, and still not being logged on the website (so no cookie or whatever), the same URL but with "abonnes" instead of "www" 

> http://**abonnes**.lemonde.fr/idees/article/2017/04/17/un_article_protege.html

registers well and save the full-length article on Wallabag!

Thus, the problem should come from the detection of the paywall and probably not the paywall and authentication bit themselves.

However, one related major drawback is that the Le Monde.fr app on Android, when using the "Share" functionality to register the article in wallabag, shares the "www.lemonde.fr" URL for all articles, even when you have a subscription. If we have in any case to modify the lemonde.fr config file, couldn't we have an option to redirect every article to "abonnes.lemonde.fr" when receiving a "lemonde.fr" or a "www.lemonde.fr" address?

Eventually, I was wondering if the registered website credentials were cleared when doing `php bin/console cache:clear -e=prod`, cause I had some problems testing with or without credentials...

Do not hesitate to ask me for more informations (I watched the debug log file but did not see anything interesting, but what do I know...).

Cheers

**Comment by techexo** [comment]
From what I understand of the configuration, there is at least two problems:

1. @etiess used `not_logged_in_xpath: //body[@class="not-logged-in"]`, which is not a body class used. I would suggest `not_logged_in_xpath: //body[@class="site-gratuit"]` as identifier.

2. More problematic, this paragraph:
> Be careful, the login form must be in the page content when wallabag loads it. It's impossible for wallabag to be authenticated on a website where the login form is loaded after the page (by ajax for example).

That is in my opinion the real issue with the Le Monde website! The form is not on the same page, but on `https://secure.lemonde.fr/sfuser/connexion?reactions=false&route=connexion&origin=origin_url.html`. Thus, it seems difficult for the code, at the moment, to work, whatever arguments we put in lemonde.fr.txt.

Workaround would be either a major improvement of the code to deal with that, or the use of the buggy website to allow the redirection of every "www.lemonde.fr" URL to "abonnes.lemonde.fr" (I am aware that it is quite borderline, however...) :)

**Comment by bdunogier** [comment]
Note that I have prototyped some stuff in regards to AJAX login forms in https://github.com/wallabag/guzzle-site-authenticator/pull/14.

**Comment by techexo** [comment]
@bdunogier Thanks for your work and your interest on the issue. I am really not confident about the arcanes of JS or AJAX and studying the source of a "teaser page" (i.e. with partial content and a link to login) I cannot see how your code would work in that case.

The link is `<a href="#" class="js_trigger_login_register" data-route="connexion">Identifiez-vous</a>`, and I don't see any corresponding script instruction elsewhere in the code.

I would be happy however to lend you my credentials (if needed) if you would be kind enough to work on this specific website! Don't hesitate to contact me :)

> After trying again, I found out that the option enabling paywall in the internal configuration wasn't even activated... Either I have some problems with my cache, either the "abonnes.lemonde.fr" website is really not that well secured (I opt for the last option, as it worked repetitively on various articles, from two internet browsers...) and Wallabag is able to save any article from the "Abonnés" version (amazing!).

I confirm that there is a serious problem with this paper's website, as the full code of the page is visible using `view-source:http://abonnes.lemonde.fr/*` URLs, and that's probably why adding directly an "abonnés" URL in wallabag works, even without credentials...

**Comment by bdunogier** [comment]
Indeed, this is interesting...  I confirm that by prepending view-source:http://abonnes... you do get the whole article. It indeed explains why Wallabag is able to get any article from the site even if the site is not configured.

This is pretty serious...

**Comment by nicosomb** [comment]
>  I confirm that by prepending view-source:http://abonnes... you do get the whole article. 

Wow!

**Comment by techexo** [comment]
I feel terribly (terribly) stupid, but I cannot find how to contact @bdunogier with Github... (yep, I'm new ;) ). Any help, please?

**Comment by j0k3r** [comment]
You can drop a message: https://gitter.im/bdunogier or https://twitter.com/bdunogier

**Comment by bdunogier** [comment]
or bertrand dot dunogier at gmail dot com, will work as well.

**Comment by techexo** [comment]
For the archives, this issue is related to @bdunogier 's [graby-site-config pull request #18](https://github.com/j0k3r/graby-site-config/pull/18) and will probably need a [custom suscriber to authorize](https://github.com/j0k3r/graby-site-config/pull/18#issuecomment-301378941) only persons with credentials.

**Comment by techexo** [comment]
Hello everyone,

I am glad to see that a pull request https://github.com/j0k3r/graby-site-config/commit/c371b0e70a74b00c28b1564e04bac69d94338c6a has been made and accepted for this issue. However, with that config file and my credentials, I still am unable to fetch full-length paywalled articles with wallabag.

Here is my debug log, trying to fetch two different articles on lemonde.fr (the first is supposed to have an url in `abonnes.lemonde.fr` once logged in, the second one in `lemonde.fr` (Le Monde Festival)): [wallabag_prod.log](https://github.com/j0k3r/graby-site-config/files/2261662/wallabag_prod.log)

As you can see, it tries to log in but at the end fetches the version "not connected". Any thoughts?

**Comment by Kdecherf** [comment]
Hello @techexo,

There's another unmerged PR required for reading paywall'd lemonde.fr articles, please see https://github.com/wallabag/guzzle-site-authenticator/pull/23

Note that this PR is on hold as I need to rework it regarding incoming changes in the project.

However you can apply the patch manually to your installation while waiting for a complete release :)

**Comment by Kdecherf** [comment]
Hello @techexo,

I saw that yesterday, it's a pretty good news actually as their change makes one of my PR obsolete. I'll work this evening to update my remaining PR.

**Comment by techexo** [comment]
If it can help, I commited (https://github.com/techexo/ftr-site-config/commit/6a0e9c45323277cc9462dcb397607b1758e933f7) an updated version of the siteconfig for `lemonde.fr.txt`, for the revamped website. I couldn't test it yet because f43.me is having some issues (https://github.com/j0k3r/f43.me/issues/193#issuecomment-438286561).

I already inserted inside the modifications made by @Kdecherf  (https://github.com/j0k3r/graby-site-config/commit/c371b0e70a74b00c28b1564e04bac69d94338c6a).

**Comment by Kdecherf** [comment]
After some checks it seems that https://github.com/fivefilters/ftr-site-config/pull/553 will be enough to fix the paywall (_read: "I will not need to patch guzzle-http-authenticator after all"_).

**Comment by Kdecherf** [comment]
@techexo thanks, I checked your changes but it appears that the `strip_class_or_id` call does not work. I moved it to a complete xpath.

**Comment by techexo** [comment]
OK, thanks for the feedback, I won't PR my changes then. I will definitely try your updated site config during the day and keep you in touch.

**Comment by Kdecherf** [comment]
@techexo my change works with the paywalled content (_I use it on my instance_) but does not bypass the restriction: you must specify credentials in order to fetch the full content.

**Comment by techexo** [comment]
@Kdecherf I have credentials, I was just wondering what were finally the modifications to merge to be able to retrieve full-length articles with wallabag.
Because if all you have to do is specify the credentials, then this issue is solved! :D

**Comment by Kdecherf** [comment]
@techexo As long as my PR in fivefilters/ftr-site-config is not released yet into wallabag, you must copy the file from master on your local installation. I'm labeling this issue as _waiting for release_.

**Comment by techexo** [comment]
@Kdecherf Thanks for your work, I can confirm I am able to retrieve `lemonde.fr` paywalled articles after pulling the last version of the site config and adding the credentials in wallabag! 👍

**Comment by Kdecherf** [comment]
Great :+1:

**Comment by Kdecherf** [comment]
Fix has been made available in the 2.3.4 release, closing this issue.
